### PR TITLE
refactor: Move CourseEditLTIFieldsEnabledFlag to lti_consumer [BD-38] [TNL-8104]

### PR DESCRIPTION
**Pull request by xitij2000**
refactor: Move CourseEditLTIFieldsEnabledFlag to lti_consumer from edx-platform.

This moves the CourseEditLTIFieldsEnabledFlag from edx-platform to this repo. This allows using the flag from both lms and studio.

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @xitij2000! I've created [OSPR-5769](https://openedx.atlassian.net/browse/OSPR-5769) to keep track of it in JIRA, where we prioritize reviews. Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.

File: lti_consumer/migrations/0012_courseeditltifieldsenabledflag.py
Comment (xitij2000) [inline lti_consumer/migrations/0012_courseeditltifieldsenabledflag.py:31]: This customisation is needed because we're moving the model from edx-platform to here, but we can't have a dependency between the migrations. For a completely new instance it's possible the table might not exist, and we'd have to create it,  However for an existing instance, the table would already exist, and Django would throw and error running the migration since the table already exists. 

So here we check if the table exists, and if it does, we simply skip creating the table _in the database_ while still creating it _in the schema_ (i.e. we are not overriding schema_forwards)..

File: lti_consumer/models.py
Comment (awaisdar001) [inline lti_consumer/models.py:561]: Using the `ConfigurationModel` as baseclass, you should be able to 
```
CourseEditLTIFieldsEnabledFlag.objects.filter(course_id=course_id).current()
```
Is there specific reason why we not using `.current()`?

File: lti_consumer/models.py
Comment (awaisdar001) [inline lti_consumer/models.py:533]: As we are moving this from edx-platform but we should name the model by convention and postfix with `Config` or `Configuration`. 
e.g. `CourseEditLTIFieldsEnabledConfig`

File: lti_consumer/models.py
Comment (awaisdar001) [inline lti_consumer/models.py:566]: you can avoid indentation by using one-liner if condition. 
```
if is_already_sharing_learner_info and not course_specific_config:
    course_config = CourseEditLTIFieldsEnabledFlag.objects.create(course_id=course_id, enabled=True)
    return course_config.enabled

```

File: lti_consumer/models.py
Comment (awaisdar001) [inline lti_consumer/models.py:574]: => `en = "" if self.enabled else en = "Not "`

Comment (awaisdar001) [reply lti_consumer/models.py:574]: I think `en` is not very readable, let's rename it to something more readable.

File: lti_consumer/plugin/views.py
Comment (awaisdar001) [inline lti_consumer/plugin/views.py:352]: Why this was removed?

File: lti_consumer/models.py
Comment (awaisdar001) [inline lti_consumer/models.py:585]: Can add a little detail about why defining `db_table` here?

File: lti_consumer/models.py
Comment (awaisdar001) [reply lti_consumer/models.py:533]: I am not sure if we can change the model name here for backward compatibility. So having that insight would be very helpful.

**Review by awaisdar001 — CHANGES_REQUESTED** [review]
Some questions and suggestions which we can improve. Otherwise looks Ok to me.

File: lti_consumer/tests/unit/test_models.py
Comment (awaisdar001) [inline lti_consumer/tests/unit/test_models.py:438]: why this needs to be indented under the previous context manager `lti_consumer_fields_editing_flag`?

**Review by awaisdar001 — CHANGES_REQUESTED** [review]

File: lti_consumer/models.py
Comment (xitij2000) [reply lti_consumer/models.py:561]: I'm not sure! I just copied it over since I din't want to change too many things at once.

File: lti_consumer/models.py
Comment (xitij2000) [reply lti_consumer/models.py:566]: Yup, I definitely leaned towards not changing any code so it would be easier to see exactly what's moving and what's added.

File: lti_consumer/models.py
Comment (xitij2000) [reply lti_consumer/models.py:585]: Sorry I did add something but it seems at some point I accidentally removed it.

File: lti_consumer/plugin/views.py
Comment (xitij2000) [reply lti_consumer/plugin/views.py:352]: pylint is complaining about a useless suppression here. Since I upgraded dependencies, the lint rules might have changed.

File: lti_consumer/tests/unit/test_models.py
Comment (xitij2000) [reply lti_consumer/tests/unit/test_models.py:438]: The first level is checking if it's true and the second is checking it with false. I don't think the indentation is necessary though.

File: lti_consumer/models.py
Comment (xitij2000) [reply lti_consumer/models.py:533]: I think we can change the name, but I feel Flag is more appropriate.  There are a couple of other ConfigurationModel-based Models that use the Flag suffix (for instance PersistentGradesEnabledFlag). 

Given that this flag will now do more than prevent editing, perhaps the name change can be more drastic. For example:
`CourseAllowLTISharePIIConfig` or `CourseEnableLTISharingPIIFlag` or something similar.

Comment (awaisdar001) [reply lti_consumer/models.py:533]: Sounds good to me - I am ok with a new pr for this change.

File: lti_consumer/tests/unit/test_models.py
Comment (awaisdar001) [reply lti_consumer/tests/unit/test_models.py:438]: Yes right.

**Review by awaisdar001 — APPROVED** [review]
@xitij2000 I am approving the PR based on our conversation on slack. 👍

File: lti_consumer/models.py
Comment (xitij2000) [reply lti_consumer/models.py:561]: I found out why this is the way it is. Calling `.current` automatically creates a new object if one doesn't exist. However in this case the object should only be created if `is_already_sharing_learner_info` is true, and not otherwise.

Comment (awaisdar001) [reply lti_consumer/models.py:561]: thanks for checking @xitij2000

Comment (awaisdar001) [reply lti_consumer/models.py:561]: thanks for checking @xitij2000

**Comment by awaisdar001** [comment]
Tagging @ormsbee for reviewing those migration changes.

File: lti_consumer/plugin/compat.py
Comment (xitij2000) [reply lti_consumer/plugin/compat.py:123]: I don't think the intention is to upgrade existing releases to this version. It's for next release of Open edX.

File: lti_consumer/models.py
Comment (xitij2000) [reply lti_consumer/models.py:581]: Sorry forgot to push the latest commit.

File: lti_consumer/migrations/0011_courseeditltifieldsenabledflag.py
Comment (xitij2000) [reply lti_consumer/migrations/0011_courseeditltifieldsenabledflag.py:39]: I didn't want to add a dependency here because that might break testing this in environments without access to edx-platform. 

It shouldn't matter what order they are run in since there are two possibilities:

1. The table already exists. In this case, the edx-platform migration have already run, and we will skip creating the table here. 
2. The table doesn't exist. In this case, the table will be created here.

File: lti_consumer/migrations/0011_courseeditltifieldsenabledflag.py
Comment (xitij2000) [reply lti_consumer/migrations/0011_courseeditltifieldsenabledflag.py:59]: The problem is that in future installations, this table will no longer exist, and it will need to be created here, so we need the database operations too. The only situation in which we don't need those operations is if the table already exists.

File: lti_consumer/plugin/compat.py
Comment (Agrendalath) [reply lti_consumer/plugin/compat.py:123]: @xitij2000, then we should consider bumping major, to indicate that it's not backward compatible change.

**Review by Agrendalath — APPROVED** [review]
@xitij2000, I'm leaving the conditional +1. Please address my comment about the versioning and add a [changelog entry](https://github.com/edx/xblock-lti-consumer#changelog).

:+1:

* [x]  I tested this: tested that this works as expected locally
* [x]  I read through the code
* [x]  I checked for accessibility issues: n/a
* [x]  Includes documentation: n/a
* [x] I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository: n/a

File: lti_consumer/plugin/compat.py
Comment (xitij2000) [reply lti_consumer/plugin/compat.py:123]: Hmm. In that case, I think I'll prefer just removing this change. It was just a drive-by cleanup. I saw this code, saw the PR was merged so decided to get rid of the TODO.

Comment (nedbat) [reply lti_consumer/plugin/compat.py]: No need to do anything for Lilac then?

Comment (Agrendalath) [reply lti_consumer/plugin/compat.py]: @xitij2000, thanks. Sure, I know this was a cleanup, though it indeed sounds like a better idea to leave this as-is for now instead of reducing the compatibility.

@nedbat, as Kshitij has removed this change from the PR, no cherry-picks are needed for Lilac here.

File: lti_consumer/plugin/compat.py
Comment (awaisdar001) [inline lti_consumer/plugin/compat.py:161]: Can you please explain why we have this import here?

**Review by awaisdar001 — APPROVED** [review]
I am generally ok with this PR but I am hoping someone from edx could take a look at the migration. Once that's done, we'd be good to go.

**Review by awaisdar001 — CHANGES_REQUESTED** [review]
Please unpin the settings in `base.in`

File: requirements/base.in
Comment (xitij2000) [reply requirements/base.in:17]: This line is copy-pasted form edx-platform: https://github.com/edx/edx-platform/blob/0be941b224f2bd0183b21be9a7b29c2666cc78ee/requirements/edx/base.in#L47

It is pinned there, and as such I pinned it here so that the version installed should be the same.

Comment (awaisdar001) [reply requirements/base.in:17]: > This line is copy-pasted form edx-platform: https://github.com/edx/edx-platform/blob/0be941b224f2bd0183b21be9a7b29c2666cc78ee/requirements/edx/base.in#L47
> 
> It is pinned there, and as such I pinned it here so that the version installed should be the same.

you are right, but OEP-18 describes something as follows. I think that needs to be updated in edx as well but because we are adding a requirement with constraint in this we should follow OEP-18. 
https://open-edx-proposals.readthedocs.io/en/latest/oep-0018-bp-python-dependencies.html#imposing-constraints-on-dependencies

File: lti_consumer/plugin/compat.py
Comment (xitij2000) [reply lti_consumer/plugin/compat.py:161]: Do you mean why it's imported at all or why it's imported inside a function?
It's being imported from the LMS to avoid duplication of this code. 
It's imported inside a function because when running tests etc we don't have edx-platform available, which will cause import errors if the import is performed at the top of the file. By putting it in a function, we avoid the import happening till the function is actually called. So during testing we can mock this function, which will avoid the import.

Comment (awaisdar001) [reply lti_consumer/plugin/compat.py:161]: Got it. :) thanks.

File: requirements/base.in
Comment (xitij2000) [reply requirements/base.in:17]: Sure, will do!

Comment (awaisdar001) [reply requirements/base.in:17]: Thanks @xitij2000

**Comment by ormsbee** [comment]
I'll look at it some more later this afternoon, but I am concerned that we're doing something fairly unusual in the migration, and I don't have a good understanding of how it will work in certain edge cases (e.g. rollback). That's not to say that this setting is done wrong, but that I need to read up more on things like `SeparateDatabaseAndState`, because I don't grok it well enough at the moment.

If I might ask some silly questions:

1. Do we need to keep the same table? Would it simplify things to just create a new model/table and copy the values over? It's < 1000 rows.
2. Is the audit history useful for this use case? Can it just be a regular model?

**Comment by xitij2000** [comment]
@ormsbee  I had no idea about the size of the table. This does indeed make it easier to copy over. However, I do feel that even in that case this migration might work better and simplify things. I hope I'm able to clarify how/why here. 

Re: SeparateDatabaseAndState
In very brief, it allows us to run a separate set of migrations on the database (via SQL) and on the state (the django model). For instance, one could drop a field in just the state, but not the database, which would make django drop the field from the model, but not remove the corresponding field from the actual database. 

Here we are deleting the model from the Django state in edx-platform, so Django no longer knows it exists, but not from the database. On the other end, in this repo, I've customised the create model operation to skip creating the table in the database, if it already exists. Also, I've made the reverse database operation a noop.

So here is what will happen on both sides of the migration:
Forwards migration:
1. The edx-platform migration runs, the model is deleted from edx-platform's record, but the data is still in the database. 
2. The xblock-lti-consumer migration runs, the model is added to the lti consumer, and uses the same table as was used by edx-platform, so all data is there.
2. If some instances of the platform are on an older version, they will keep being able to use that model, and read data from it because it exists in the same place. 
3. If some instances are on the newer version, they will be able to read data from the same table, but the model will just be loaded from lti consumer. 
4. Any data either the new version or the old version write to the table will be available to both instances since it's the same table with the same format. 

Reverse migration:
1. The lti consumer migration runs, the model is deleted from its record, but the data is still in the database since the reverse database migration is a noop.
2. The edx-platform migration runs, the model is recreated in django's record, but not n the database since we used a SeparateDatabaseAndState migration that only runs for the state. 
3. The data for both is in the same table, the same location so newer and older versions can still continue to read/write the same data.

**Comment by awaisdar001** [comment]
> I'll look at it some more later this afternoon, but I am concerned that we're doing something fairly unusual in the migration, and I don't have a good understanding of how it will work in certain edge cases (e.g. rollback). That's not to say that this is being done wrong, but that I need to read up more on things like `SeparateDatabaseAndState`, because I don't grok it well enough at the moment.
> 
> If I might ask some silly questions:
> 
> 1. Do we need to keep the same table? Would it simplify things to just create a new model/table and copy the values over? It's < 1000 rows.
> 2. Is the audit history useful for this use case? Can it just be a regular model?

I can add some context about the size of the table. 

Prod-Edxapp: 874 rows.
Edge-edxapp: 96 rows.

**Review by ormsbee — APPROVED** [review]

File: setup.py
Comment (nedbat) [inline setup.py:52]: A version bump deserves an entry in the README history :)

**Comment by nedbat** [comment]
This will need a rebase now.

**Comment by nedbat** [comment]
Also, the version number will have to become 2.10.0

File: setup.py
Comment (xitij2000) [reply setup.py:52]: Sure, will do!

**Comment by awaisdar001** [comment]
@xitij2000 there are some quality tests failing on this PR. 
Do you have access to merge this PR?

**Comment by awaisdar001** [comment]
> @awaisdar001 It seems a previous PR didn't include documentation about a toggle. Not sure why it wasn't flagged then.
> 
> I'll see how I can fix this.

thanks @xitij2000

**Comment by awaisdar001** [comment]
> @awaisdar001 It seems a previous PR didn't include documentation about a toggle. Not sure why it wasn't flagged then.
> 
> I'll see how I can fix this.

@xitij2000 try adding `pylint: disable=toggle-missing-annotation` to remove the error if this is not caused by your change.

**Review by giovannicimolin — COMMENTED** [review]
@xitij2000 @nedbat merged https://github.com/edx/xblock-lti-consumer/pull/166 with the flag annotation fix.

Can you replace the usage of https://github.com/edx/xblock-lti-consumer/blob/master/lti_consumer/utils.py#L23 with this new flag implemented here? 

I've already updated the ADR with the change (see https://github.com/edx/xblock-lti-consumer/blob/master/docs/decisions/0004-lti-advantage-nrps.rst#status)

**Comment by awaisdar001** [comment]
> @xitij2000 @nedbat merged #166 with the flag annotation fix.
> 
> Can you replace the usage of https://github.com/edx/xblock-lti-consumer/blob/master/lti_consumer/utils.py#L23 with this new flag implemented here?
> 
> I've already updated the ADR with the change (see https://github.com/edx/xblock-lti-consumer/blob/master/docs/decisions/0004-lti-advantage-nrps.rst#status)

also @xitij2000 if the waffle is related to this project can you please also add that here?
https://openedx.atlassian.net/wiki/spaces/COMM/pages/2884665541/BD-38+Flags+Toggles

**Review by awaisdar001 — APPROVED** [review]

**Comment by xitij2000** [comment]
@giovannicimolin  There are many changes that need to be applied to this PR, but I wanted to keep this PR a simple transplant of edx-platform code to xblock-lti-consumer, so I'm avoiding those changes. In the next PR (coming shortly) I will clean up more of this code and apply the change you suggest. 

@awaisdar001 This is not a new flag being added, but it is relevant to the project, so I'll add it there.

**Comment by awaisdar001** [comment]
> @giovannicimolin There are many changes that need to be applied to this PR, but I wanted to keep this PR a simple transplant of edx-platform code to xblock-lti-consumer, so I'm avoiding those changes. In the next PR (coming shortly) I will clean up more of this code and apply the change you suggest.
> 
> @awaisdar001 This is not a new flag being added, but it is relevant to the project, so I'll add it there.

@xitij2000 yes -- we'd like to add any waffles related to this project. Thank you.

**Comment by awaisdar001** [comment]
@xitij2000 This PR is done from our side. I have updated the status in the confluence doc. If you need help merging this PR until you get merge access back, please let me know when it's ready to merge from your side.

**Comment by awaisdar001** [comment]
PR merged for @xitij2000 as he do not have access to merge.

**Comment by nedbat** [comment]
Doesn't this pull request have to be carefully coordinated with https://github.com/edx/edx-platform/pull/27529?  We shouldn't have a release of this repo before that change in edx-platform is merged, right?

**Comment by awaisdar001** [comment]
Thanks for brining it up @nedbat - @xitij2000 needed the new version for the PR referenced above. 

I think for now what we can do is to add a constraint in the requirements and it will stop the error from surfacing. I'll keep @xitij2000 in loop when he is around. 

Thoughts about pining the version a constraints file?

**Comment by nedbat** [comment]
Yes, it looks like we should pin the version for now.  edx-platform names this unpinned in its base.in file.

**Comment by nedbat** [comment]
And actually, @xitij2000, it seems like we need to think carefully about how to manage this change beyond edx.org also.  This new version of the block won't work on Lilac unless we also cherry-pick the edx-platform pull request.
Should the major version be bumped because of the incompatibility?

**Comment by nedbat** [comment]
Diana made a pull request: https://github.com/edx/edx-platform/pull/28014

**Comment by xitij2000** [comment]
> Doesn't this pull request have to be carefully coordinated with [edx/edx-platform#27529](https://github.com/edx/edx-platform/pull/27529)? We shouldn't have a release of this repo before that change in edx-platform is merged, right?

Yes, I think that PR should have been merged soon after 2.11 was tagged and released. 

> And actually, @xitij2000, it seems like we need to think carefully about how to manage this change beyond edx.org also. This new version of the block won't work on Lilac unless we also cherry-pick the edx-platform pull request.
> Should the major version be bumped because of the incompatibility?

I have another change in the pipeline that renames the affected model based on the discussion [here](https://github.com/edx/xblock-lti-consumer/pull/159#discussion_r627158475), so I can update that PR to bump up the major version. 

Since 2.11 is already out there, perhaps I can update the changelog to indicate that 2.11 doesn't work with Lilac and earlier? If there are other changes lined up in this repo that need to got to Lilac then perhaps cherry-picking the change would be more convenient?

**Pull request merged**
